python/ray/data/dataset.py

```diff
@@ -292,13 +296,21 @@ def transform(block: Block) -> Iterable[Block]:
                 # bug where we include the entire base view on serialization.
                 view = block.slice(start, end, copy=batch_size is not None)
                 if batch_format == "native":
-                    # Always promote Arrow blocks to pandas for consistency.
-                    if isinstance(view, pa.Table) or isinstance(view, bytes):
+                    if isinstance(view, pa.Table):
+                        if view.column_names == [TENSOR_COL_NAME]:
+                            view = BlockAccessor.for_block(view).to_numpy()
```

Comment:
This doesn't seem safe. What if I had a non-numpy Arrow table with "value" as a field?

It would be preferable to make the rule something like: we always convert if there is a single column, and that column is a numpy array.

Separately, I'm also wondering if we should get rid of "simple" block format and treat that as some kind of single-column Arrow table as well, but that can be a future issue.

Reply:
>This doesn't seem safe. What if I had a non-numpy Arrow table with "value" as a field?

That's why:
1. I made the dummy tensor column name an internal column name, `"__RAY_TC__"`, that shouldn't be accidentally collided with like `"value"`.
2. I'm creating a strong separation between the representations of the pure-tensor dataset single-column case, `"__RAY_TC__"`, and our creation of single column tables via `range_table()`, `"value"`, since we know that the former is always cleanly convertible to an ndarray, while we do not know if the latter is.

>It would be preferable to make the rule something like: we always convert if there is a single column, and that column is a numpy array.

Hmm I don't know if that is a good idea; this automatic conversion makes sense for the pure-tensor dataset case (created from `range_tensor()` or `from_numpy()`) since the single-column table representation is a Datasets implementation detail, while the model in their mind is a tensor.

Meanwhile, the user could be surprised when their single-column table **that they created themselves** is transformed into an ndarray without them asking for such a conversion:
```python
df = pd.DataFrame({"img": TensorArray(...)})
ds = ray.data.from_pandas(df)

def mapper(df: pd.DataFrame):
    # Add new derived columns
    df["img2"] = df["img"] + 1
    df["mean_pixel"] = df["img"].to_numpy().mean(axis=(1, 2))
    return df

# This will fail if the batches are automatically turned into an ndarray.
ds = ray.data.map_batches(mapper)
```

>Separately, I'm also wondering if we should get rid of "simple" block format and treat that as some kind of single-column Arrow table as well, but that can be a future issue.

I was recently thinking about this is an optimization for other types, e.g. rows of strings, which we could represent as a single-column Arrow block, but I think that we'll always want the simple block as an escape hatch for types that we can't represent in Arrow, unless the maintenance or feature constraint burden for simple blocks gets too high.

Reply:
We're representing a tensor block as single-tensor-column table
```python
TensorBlock = TabularBlock[
    columns=[
        TabularColumn[name="__RAY_TC__", type=TensorExtensionType]
    ]
]
```
and in this PR, we're converting between this internal representation and the equivalent NumPy ndarray at the boundaries of the UDF and in the consuming iterators. We can precisely identify a tensor block and do this well-defined/efficient conversion by checking for a single `"__RAY_TC__"` column in tabular blocks, while just checking for a single-column tensor table could run into the unexpected automatic conversion example given above.

Reply:
I think something like this makes sense, but I'd like to have a unified plan for how this will look into the future. It would be really great if we could get rid of most of our custom block support, and only have ArrowBlock + some translation layers for the tensor and simple python cases.